### PR TITLE
feat: add options to allow specific routes helmet configuration and harmonize repository structure

### LINKED ISSUE #161: How to disable helmet contentSecurityPolicy for specific reply? Or disable helmet?
## 💬  Question here 

Hi, need to disable helmet contentSecurityPolicy for a few specific replies:
```
    if (someCondition) { disable contentSecurityPolicy }
    reply.type('text/html') //otherwise it's plain text, instead of being interpreted as html
    reply.send(page) //this has a <script> which doesn't execute due to contentSecurityPolicy
```

helmet is added with `register()`, part of app.js, launched using fastify-cli
```js
 fastify.register(helmet)
```

Found these, but didn't figure out a solution:
https://github.com/helmetjs/helmet/wiki/Conditionally-using-middleware

https://www.fastify.io/docs/latest/Reference/Middleware/#restrict-middleware-execution-to-certain-paths
(not using fastify.use(), but  fastify.register())

https://github.com/fastify/fastify-helmet/issues/89

Tried also
```js
//still uses helmet
    reply.raw.writeHead(200, { 'Content-Type': 'text/html' }) 
    reply.raw.write(page)
     reply.raw.end()
```
Thanks!

- *node version*: 16
- *fastify version*: 3
- *fastify-helmet*: 4

### PR DESCRIPTION
**Pull request by darkgl0w**
Hello.

This PR aims to:
- add options to allow specific routes helmet configuration (closes https://github.com/fastify/fastify-helmet/issues/161)
- update examples to actually use the current plugin code with the new feature
- add a `reply.helmet` decorator
- harmonize the repository structure
- add `.taprc` configuration file
- upgrade package dependencies
- update license date to `2017-2022`
- update typescript definitions and typescript definitions tests
- document the new `reply.helmet` decorator, the new `global` configuration option and the new `helmet` shorthand route option
- cover more route specific use cases inside the `routes.test.js` file

> NB: This PR does not introduce any breaking change and can land as a minor.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — COMMENTED** [review]
Good work!

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by darkgl0w** [comment]
The PR is ready to receive reviews.
I have been careful but I'm a little tired so there may be some typos, and I am not sure about how I named the different variables and functions, suggestions are welcomed :stuck_out_tongue:.

File: README.md [hunk 1]
```diff
@@ -11,7 +11,7 @@ Important security headers for Fastify. It is a tiny wrapper around
 
 ## Install
 ```
-npm i fastify-helmet
+npm install --save fastify-helmet
```
Comment (Fdawgs) [inline README.md:14]:
```suggestion
npm i fastify-helmet
```

Isn't "--save" redundant as of npm v5?

Comment (darkgl0w) [reply README.md]: Oh right, If you didn't point that fact I have my habits with `-s` and `--save` when I install a package that I would have never realized that it now automatically save the installed package in the lock file :stuck_out_tongue:

File: README.md [hunk 2]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
 })
 ```
 
+## How it works
+
+`fastify-helmet` is just a tiny wrapper around helmet that adds an `'onRequest'` hook
```
Comment (Fdawgs) [inline README.md:38]:
```suggestion
`fastify-helmet` is a tiny wrapper around helmet that adds an `'onRequest'` hook
```

File: README.md [hunk 3]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+and a `reply.helmet` decorator.
+
+It accepts the same options than Helmet, and you can see more in [the helmet documentation](https://helmetjs.github.io/docs/).
```
Comment (Fdawgs) [inline README.md:41]:
```suggestion
It accepts the same options as helmet, and you can see more in [the helmet documentation](https://helmetjs.github.io/docs/).
```

File: README.md [hunk 4]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+  }
+}, async (request, reply) => {
+  return { message: 'helmet is enabled here' }
+})
+
+// helmet is disabled on this route but we have access to `reply.helmet` decorator
+// that allows us to apply helmet conditionally
+fastify.get('/here-we-use-helmet-reply-decorator', async (request, reply) => {
+  if (condition) {
+    // we apply the defaut options
```
Comment (Fdawgs) [inline README.md:91]:
```suggestion
    // we apply the default options
```

File: README.md [hunk 5]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+
+  return { 
+    message: 'we use the helmet reply decorator to conditionally apply helmet middlewares'
+  }
+})
+```
+
+### `helmet` route option
+
+`fastify-helmet` allows you to enable, disable and customize helmet for each one of your application hook by using the
```
Comment (Fdawgs) [inline README.md:106]:
```suggestion
`fastify-helmet` allows you to enable, disable, and customize helmet for each one of your application hook by using the 
```

**Review by Fdawgs — COMMENTED** [review]
Will take a longer pass tomorrow (UTC time) if that's ok? So far so good from what I've seen, good work!

File: README.md [hunk 6]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+`fastify-helmet` is a tiny wrapper around helmet that adds an `'onRequest'` hook
+and a `reply.helmet` decorator.
+
+It accepts the same options as helmet, and you can see more in [the helmet documentation](https://helmetjs.github.io/docs/).
+
+### Apply Helmet to all your application routes
+
+By passing `{ global: true }` into the options, `fastify-helmet` allows you to register Helmet for all your application
+routes by default. If you want a more granular control on how to apply Helmet to your application you can choose to
+disable it on a global scope by passing `{ global: false }` to the options. By default this option is set to `true`.
```
Comment (Fdawgs) [inline README.md:47]:
```suggestion
disable it on a global scope by passing `{ global: false }` to the options. By default, this option is set to `true`.
```

File: README.md [hunk 7]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+
+  return { 
+    message: 'we use the helmet reply decorator to conditionally apply helmet middlewares'
+  }
+})
+```
+
+### `helmet` route option
+
+`fastify-helmet` allows you to enable, disable, and customize helmet for each one of your application hook by using the
```
Comment (Fdawgs) [inline README.md:106]:
```suggestion
`fastify-helmet` allows you to enable, disable, and customize helmet for each one of your application hooks by using the 
```

File: README.md [hunk 8]
```diff
@@ -33,15 +33,141 @@ fastify.listen(3000, err => {
...
+```
+
+### `helmet` route option
+
+`fastify-helmet` allows you to enable, disable, and customize helmet for each one of your application hook by using the 
+`helmet` shorthand route option when you register your application routes.
+
+If you want to disable helmet for a specific endpoint you must pass `{ helmet: false }` to your route options.
+
+If you want to enable or customize helmet for a specific endpoint you must pass an helmet configuration object to your
```
Comment (Fdawgs) [inline README.md:111]:
```suggestion
If you want to enable or customize helmet for a specific endpoint you must pass a helmet configuration object to your
```

File: index.js [hunk 9]
```diff
@@ -1,53 +1,131 @@
...
 const fp = require('fastify-plugin')
 const helmet = require('helmet')
-const crypto = require('crypto')
 
-module.exports = fp(function (app, options, next) {
-  const enableCSPNonces = options.enableCSPNonces
-  // clear options as helmet will throw when any options is "true"
-  options.enableCSPNonces = undefined
+function helmetPlugin (fastify, options, next) {
+  // helmet will throw when any options is explicitly set to "true"
```
Comment (Fdawgs) [inline index.js:8]:
```suggestion
  // helmet will throw when any option is explicitly set to "true"
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]
A few minor spelling/grammar mistakes that you had already alluded to but all good otherwise!

File: index.js (hunk 9 above)
Comment (darkgl0w) [reply index.js]: Applied all your suggestions.
Thanks a lot for taking the time to review this PR ^^

**Review by Fdawgs — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Review by climba03003 — APPROVED** [review]
@mcollina 
As it is a heavy refactor on project structure, should it be a `major` bump to prevent any potential breakage that we didn't found?

The new structure is good to me.

**Comment by darkgl0w** [comment]
> @mcollina As it is a heavy refactor on project structure, should it be a `major` bump to prevent any potential breakage that we didn't found?
> 
> The new structure is good to me.

@climba03003 > This PR branch code can be dropped in on current code bases and work exactly like the the previous version of the package (the default values of the newly introduced options mimic the way this package currently work).
- One of the major requirements in the roadmap I followed was to completely pass the old test suite (the master branch `test.js` file) without modifying a single character in the test file 😛. Running the `test.js` suite of the master branch against this PR will pass all the tests.

- the `global.test.js` file is mostly the old `test.js` file with a little code refresh and additional tests scenarii where the new features are used.

I have been extremely careful when implementing these features to not introduce any breaking change in order to land this PR as a `minor` but as a precaution we can land this as a `major`.

**Comment by mcollina** [comment]
I'd go with a major as a safety precaution, integers are cheap.

**Comment by Fdawgs** [comment]
It was a good idea with the major bump. With my APIs i would register helmet at a global level and then child contexts would have their own plugins and configs, so the ones hosting the docs pages would have relaxed CSPs. This now no longer works (I was probably misusing the plugin to begin with).

Example:
```js

const fastify = require("fastify")();
const helmet = require("fastify-helmet");

async function server() {
	fastify
		.register(helmet)
		.route({
			method: "GET",
			url: "/foo",
			handler: (req, res) => {
				res.send("ok");
			},
		})
		.register(async (childContext) => {
			childContext
				.register(helmet, {
					contentSecurityPolicy: {
						directives: {
							"script-src": ["'self'", "'unsafe-inline'"],
							"style-src": ["'self'", "'unsafe-inline'"],
							"child-src": ["'self'", "blob:"],
						},
					},
				})
				.route({
					method: "GET",
					url: "/child",
					handler: (req, res) => {
						res.send("fake docs here");
					},
				});
		});

	await fastify.listen({ port: 8000 });
}

server();
```

This now throws: `FastifyError [Error]: The decorator 'helmet' has already been added!`

**Comment by darkgl0w** [comment]
@Fdawgs > Hmmm good spot !
My bad I should have first checked if helmet decorator was already defined before trying to decorate one.

I will add a fix and a proper test for this use case.

**Comment by darkgl0w** [comment]
@Fdawgs > PR submitted it should cover your use case ^^

**Pull request merged**